etcdserver/cindex/cindex.go

```diff
@@ -0,0 +1,84 @@
+// Copyright 2015 The etcd Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cindex
+
+import (
+	"encoding/binary"
+	"sync/atomic"
+
+	"go.etcd.io/etcd/mvcc/backend"
+)
+
+var (
+	metaBucketName = []byte("meta")
+
+	consistentIndexKeyName = []byte("consistent_index")
+)
+
+// ConsistentIndexer is an interface that wraps the Get/Save method.
+// Consistent index is the offset of an entry in a consistent replicated log.
+type ConsistentIndexer interface {
+	// ConsistentIndex returns the consistent index of current executing entry.
+	ConsistentIndex() uint64
+
+	// Save saves consistentIndex to the underlying stable storage.
+	Save(tx backend.BatchTx)
+}
+
+// consistentIndex represents the offset of an entry in a consistent replica log.
+// It implements the ConsistentIndexer interface.
+// It is always set to the offset of current entry before executing the entry,
+// so ConsistentWatchableKV could get the consistent index from it.
+
+type ConsistentIndex struct {
+	consistentIndex uint64
+	bytesBuf8       []byte
+}
+
+func NewConsistentIndex() ConsistentIndex {
+	return ConsistentIndex{bytesBuf8: make([]byte, 8)}
+}
+
+func (cindex *ConsistentIndex) LoadConsistentIndex(tx backend.BatchTx) uint64 {
+
+	if ci := atomic.LoadUint64(&cindex.consistentIndex); ci > 0 {
+		return ci
+	}
+	tx.Lock()
+	defer tx.Unlock()
+	_, vs := tx.UnsafeRange(metaBucketName, consistentIndexKeyName, nil, 0)
+	if len(vs) == 0 {
+		return 0
+	}
+	v := binary.BigEndian.Uint64(vs[0])
+	atomic.StoreUint64(&cindex.consistentIndex, v)
+	return v
+}
+
+func (cindex *ConsistentIndex) ConsistentIndex() uint64 {
+	return atomic.LoadUint64(&cindex.consistentIndex)
+}
+
+func (cindex *ConsistentIndex) SetConsistentIndex(v uint64) {
+	atomic.StoreUint64(&cindex.consistentIndex, v)
+}
+
+func (cindex *ConsistentIndex) Save(tx backend.BatchTx) {
```

Comment:
This function dose not hold lock, probably should call this function `UnsafeSave`.

Optionally, we could have an actual `Save` function which acquires lock before saving the index. The `Save` function is currently not used by etcd code, just for completeness.

Reply:
updated. thanks.